module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+                // The initial root hashes don't match, just slash the attester
+                return Err(self.slash_burn_reward(
+                    attester,
+                    Role::Attester,
+                    SlashingReason::InvalidInitialHash,
+                    working_set,
+                ));
+            }
+        } else {
+            // Genesis state
+            // We can assume that the genesis hash is always set, otherwise we need to panic.
+            // We don't need to prove that the attester was bonded, simply need to check that the current bond is higher than the
+            // minimal bond and that the attester is not unbonding
+            if self
+                .chain_state
+                .get_genesis_hash(working_set)
+                .expect("The initial hash should be set")
+                != attestation.initial_state_root
+            {
+                // Slash the attester, and burn the fees
+                return Err(self.slash_burn_reward(
+                    attester,
+                    Role::Attester,
+                    SlashingReason::InvalidInitialHash,
+                    working_set,
+                ));
+            }
+            // Normal state
+        }
+
+        Ok(CallResponse::default())
+    }
+
+    /// Try to process an attestation, if the attester is bonded
+    pub(crate) fn process_attestation(
+        &self,
+        context: &C,
+        attestation: Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // We first need to check that the attester is still in the bonding set
+        if self
+            .bonded_attesters
+            .get(context.sender(), working_set)
+            .is_none()
+        {
+            return Err(AttesterIncentiveErrors::UserNotBonded);
+        }
+
+        // We first need to check the bonding proof (because we can't slash the attester if he's not bonded)
+        self.check_bonding_proof(context, &attestation, working_set)?;
+
+        // We suppose that these values are always defined, otherwise we panic
+        let last_attested_height = self
+            .maximum_attested_height
+            .get(working_set)
+            .expect("The maximum attested height should be set at genesis");
+        let current_finalized_height = self
+            .light_client_finalized_height
+            .get(working_set)
+            .expect("The light client finalized height should be set at genesis");
+        let finality = self
+            .rollup_finality_period
+            .get(working_set)
+            .expect("The rollup finality period should be set at genesis");
+
+        // Update the max_attested_height in case the blocks have already been finalized
+        let new_height_to_attest = max(last_attested_height, current_finalized_height) + 1;
```

Comment:
Can the `last_attested_height` ever be less than the `current_finalized_height`? If so, that seems like a cause for concern - light clients will have accepted a state transition which was never attested to!

Reply:
That was meant to deal with the case where the state gets updated without any attestation being posted during the finalizing period. But I think now I remember you saying that we have to revert in that case.

Reply:
Ok, so I will add a panic and remove the max